contract_payment_mode/models/contract.py

```diff
@@ -0,0 +1,27 @@
+from odoo import api, fields, models
+
+
+class ContractContract(models.Model):
+    _inherit = "contract.contract"
+
+    payment_mode_id = fields.Many2one(
+        comodel_name="account.payment.mode",
+        string="Payment Mode",
+        domain=[("payment_type", "=", "inbound")],
+        index=True,
+    )
+
+    @api.onchange("partner_id")
+    def on_change_partner_id(self):
+        self.payment_mode_id = self.partner_id.customer_payment_mode_id.id
+
+    def _prepare_invoice(self, date_invoice, journal=None):
+        res = super(ContractContract, self)._prepare_invoice(
+            date_invoice=date_invoice, journal=journal
+        )
+        invoice_vals = res[0]
+        if self.payment_mode_id:
+            invoice_vals["payment_mode_id"] = self.payment_mode_id.id
+            invoice = self.env["account.move"].new(invoice_vals)
+            invoice_vals = invoice._convert_to_write(invoice._cache)
```

Comment:
@pedrobaeza come with the idea to use Form test class to trigger the onchanges. Can you do the same here.

Please check

https://github.com/OCA/contract/blob/6177f9de43a7acd26bca05e7dc905f7c9622f07a/contract/models/contract.py#L322-L344

Reply:
Yes, the idea is to reuse `move_form` variable if needed to perform changes with onchanges, or directly inject in `invoice_vals` any value in the dictionary. That way, you have the flexibility of using any of this mechanism, but you should respect at least the expected return value.

Reply:
They way I've done it I have to pop these values or else they get overwritten:
https://github.com/OCA/contract/blob/6177f9de43a7acd26bca05e7dc905f7c9622f07a/contract/models/contract.py#L333-L342

Another way of doing this could be to have an extra function in contract/contract/models/contract.py to modify the fields of `move_form` so I only have to inherit that function without having to call `_values_to_save()`.